### PR TITLE
Update injector architecture.

### LINKED ISSUE #89: Modify injector architecture
**Describe the task**
It seems that [javaparser](http://javaparser.org) is robust at parsing java source code. Preserving the code style is really important for our use case, considering that we had some issues with annotation insertion / deletion we might be able to utilize javaparser in computing the offsets where the `@Nullable` annotation should be added and then directly modify the content of the containing source file.

**Expected behavior**
All the annotations get inserted without breaking anything in the code. (modifier removal, syntax errors or changing the code style).

**Additional context**
With this issue closed, #77 and #78 can also be closed.

### PR DESCRIPTION
**Pull request by nimakarimipour**
This PR resolves #89 by computing the offsets where an annotation should be added using javaparser and directly inserting the annotation as string at the corresponding offsets in the source file.

Changes in the AST such as `AddAnnotation` / `RemoveAnnotation` are translated into text modification instances (`Insertion`, `Deletion` or `Replacement`) and applied to source file using the printer.

Imports are also added using the same technique above.
To correctly compute the offset where the new import declarations should be added we follow the approach below:
At each step, if the desired location is not found, it goes to the following step.
1. Immediately after the last existing import statement
2. Immediately after the package declaration statement
3. Immediately after the copy right header
4. Immediately before non-empty line

### CODE REVIEW

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        && !Character.isLetterOrDigit(line.charAt(head))
+        && line.charAt(head + 1) == ' ') {
+      // remove extra white space
+      line.deleteCharAt(head + 1);
+    }
+    String removed = line.toString();
+    if (removed.strip().equals("")) {
+      lines.remove(startPosition.line);
+      return;
+    }
```
Comment (msridhar) [inline injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java:67]: Do you have this because Google Java Format won't delete a blank line or something?  I'd prefer to just let GJF do the work it does to minimize the complexity of our code

File: injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java [hunk 2]
```diff
@@ -58,8 +62,8 @@ protected void fillJsonInformation(JSONObject res) {
   }
 
   @Override
-  protected boolean applyToMember(NodeList<BodyDeclaration<?>> clazz, Change change) {
-    final boolean[] success = {false};
+  protected Modification applyToMember(NodeList<BodyDeclaration<?>> clazz, Change change) {
+    final Modification[] modification = new Modification[1];
```
Comment (msridhar) [inline injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java:66]: Is this an array just to make it mutable?  If so, I think a more common pattern is to use `AtomicReference` for this purpose

File: injector/src/main/java/edu/ucr/cs/riple/injector/location/OnMethod.java [hunk 3]
```diff
@@ -55,31 +59,38 @@ protected void fillJsonInformation(JSONObject res) {
   }
 
   @Override
-  protected boolean applyToMember(NodeList<BodyDeclaration<?>> clazz, Change change) {
-    final boolean[] success = {false};
+  protected Modification applyToMember(NodeList<BodyDeclaration<?>> clazz, Change change) {
+    final Modification[] modification = new Modification[1];
```
Comment (msridhar) [inline injector/src/main/java/edu/ucr/cs/riple/injector/location/OnMethod.java:63]: Same array comment here

File: injector/src/main/java/edu/ucr/cs/riple/injector/location/OnParameter.java [hunk 4]
```diff
@@ -59,8 +64,8 @@ protected void fillJsonInformation(JSONObject res) {
   }
 
   @Override
-  protected boolean applyToMember(NodeList<BodyDeclaration<?>> members, Change change) {
-    final boolean[] success = {false};
+  protected Modification applyToMember(NodeList<BodyDeclaration<?>> members, Change change) {
+    final Modification[] modification = new Modification[1];
```
Comment (msridhar) [inline injector/src/main/java/edu/ucr/cs/riple/injector/location/OnParameter.java:68]: Array comment again

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Insertion.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+public class Insertion extends Modification {
+
+  public Insertion(String content, Position position, ElementKind kind) {
+    super(content, position, kind);
+  }
+
+  @Override
+  public void visit(List<String> lines) {
+    String toInsert;
+    String line = lines.get(startPosition.line);
```
Comment (msridhar) [inline injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Insertion.java:41]: As a high-level comment, some parsers provide character offsets for each AST node in addition to line+column.  In an approach based on character offsets, you may not need to worry about adding / deleting lines as much.  But I don't know if JavaParser gives you that info, and the line-based approach might be more efficient.  No need to change everything now just wanted to jot the thought

**Review by msridhar — APPROVED** [review]
This looks quite good to me, but I am not an expert in this code so I may be missing something.  Approving since it seems well-tested and with this new approach we should be able to fix bugs that arise much more easily.  Comments are minor

File: injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java (hunk 2 above)
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java]: Yes, exactly. Thank you for the suggestion, changed it 15586f197edd93b52343e635c4386601abe1ee5b.

File: injector/src/main/java/edu/ucr/cs/riple/injector/location/OnMethod.java (hunk 3 above)
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/location/OnMethod.java]: ae5fa24695e5a4317d40d5dba69eacb6f7cbd66e

File: injector/src/main/java/edu/ucr/cs/riple/injector/location/OnParameter.java (hunk 4 above)
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/location/OnParameter.java]: 9e41e3f59ad330db20417a85e729a25a9cfe469f

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  public Deletion(String content, Position startPosition, Position endPosition, ElementKind kind) {
+    super(content, startPosition, kind);
+    this.endPosition = new Position(endPosition.line - 1, endPosition.column - 1);
+  }
+
+  @Override
+  public void visit(List<String> lines) {
+    int start = startPosition.column;
+    int end = endPosition.column;
+    StringBuilder line = new StringBuilder(lines.get(startPosition.line));
```
Comment (lazaroclapp) [inline injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java:46]: This assumes `startPosition.line == endPosition.line` for deletions, no? Meaning that the entire annotation should be within the same line.

1. We probably should make the assumption explicit in comments and maybe a preconditions check
2. This might get broken for the case of `@SuppressWarnings` annotations, eg.

```
@SuppressWarnings(
    "NullAway.Init")
@SomeOtherAnnotation(...)
private Object f1;
```

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Replacement.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    super(content, startPosition, kind);
+    this.endPosition = new Position(endPosition.line - 1, endPosition.column - 1);
+    if (content.equals("")) {
+      throw new IllegalArgumentException("content cannot be empty, use Deletion instead");
+    }
+  }
+
+  @Override
+  public void visit(List<String> lines) {
+    StringBuilder line = new StringBuilder(lines.get(startPosition.line));
```
Comment (lazaroclapp) [inline injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Replacement.java:48]: This assumes that the thing being replaced is within a single line. This is even more unlikely when removing an argument for an annotation, consider:

```
@SuppressWarnings(
   "InvalidThrowsLink",
   "MalformedInlineTag",
   "NullAway",
   "unchecked"
)
private void foo();
```

**Review by lazaroclapp — COMMENTED** [review]
Did a pass. I do think there are some comments and edge-cases worth taking a look at...

File: injector/src/test/java/edu/ucr/cs/riple/injector/AnnotationWithArgumentTest.java [hunk 8]
```diff
@@ -237,7 +238,6 @@ public void onMethodCustomAnnotMultiArgExistsTest() {
             "}")
         .expectOutput(
             "package com.edu;",
-            "import edu.ucr.CustomNull;",
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/test/java/edu/ucr/cs/riple/injector/AnnotationWithArgumentTest.java]: Thank you very much for noticing this. Actually it should be present in both cases, the reason that we do not see the import added in the output is that the exact annotation is already present, and annotator does not do anything in this case. And to answer to your second question, No, we do not compile any codes in tests, we just parse the codes. 7aebb6b314adc5c1641cd1833eecf07064ea585a

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java (hunk 1 above)
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java:67]: Yes exactly, if we don't remove empty lines, Annotator will leave the source files with a lot of empty lines added as annotations are added and then deleted later. Also, annotator only requests deletion of annotations that it injected itself. As we know the added annotations are always followed by a new line, this code is mostly safe.

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Insertion.java (hunk 5 above)
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Insertion.java:41]: Thank you for the suggestion, I did not see this structure in javaparser, only the `Position` instance is available which keeps track of line and column.

File: injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+            "}")
+        .expectOutput(
+            "package edu.ucr;",
+            "import javax.annotation.Nullable;",
+            "public class Test {",
+            "   Field foo1;",
+            "   @Nullable",
+            "   Field foo2;",
+            "   @Nullable",
+            "   Field foo3, foo4;",
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java:74]: This is regarding the decision we made long ago, as it is not always trivial how to break multiple inline field declarations into single declarations, any annotation on any field in a group is added for all of them. This is also considered in collecting the potential impacted regions. In this case the potential impacted regions of `foo3` is the union of the impacted regions of `foo3` and `foo4`.

File: injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+            "       return new Object();",
+            "   }",
+            "}")
+        .expectOutput(
+            "package edu.ucr;",
+            "import javax.annotation.Nullable;", // We do not support removing imports
+            "public class Test {",
+            "   Field foo1;",
+            "   Field foo2;",
+            "   Field foo3, foo4;",
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java:163]: In the view of annotator, there is no fix targeting just element `foo4` or `foo3`. The target location is the group of `foo3` and `foo4`. All serialized fixes from NullAway is also adapted to this approach.

File: injector/src/main/java/edu/ucr/cs/riple/injector/Printer.java [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+   */
+  public void applyModifications(Set<Modification> modifications) {
+    // Modification are sorted to start from the last position, to make the changes ineffective to
+    // current computed offsets.
+    modifications.stream()
+        .sorted(
+            Comparator.comparingInt((Modification o) -> o.startPosition.line)
+                .thenComparingInt(o -> o.startPosition.column)
+                .reversed())
+        .forEach(modification -> modification.visit(lines));
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/Printer.java:75]: This is a great point and we should definitely document this limitation. We do not support addition and deletion within one run of Injector. I rewrote `injector` based on the use case of the `core` module and in that use case we only request additions or deletions per run of injector. This limitation was resolved in core since we have apis with name `injectFixes` or `removeFixes`. I updated them to reflect this constraint even further. And we do support addition of multiple annotations on single field, added a unit test for that.

3eade0d2d1d3054f73c6835b68077c6c563ea264
e8d887ed529e4a9fc366e6c7f7be6cf61efd69eb
0c9ad7fc897f6f0cf0950d57dee34d88a27012b7

File: injector/src/main/java/edu/ucr/cs/riple/injector/Printer.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+  /**
+   * Computes the offset where the import declarations should be added. Below is the steps it takes
+   * to compute the correct offset. At each step, if the desired location is not found, it goes to
+   * following step.
+   *
+   * <ol>
+   *   <li>Immediately after the last existing import statement
+   *   <li>Immediately after the package declaration statement
+   *   <li>Immediately after the copy right header
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/Printer.java]: d00be48cc9b7a65723a4e03a155fe44d895d7120

File: injector/src/main/java/edu/ucr/cs/riple/injector/Printer.java [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+      Optional<Range> range = optional.get().getRange();
+      if (range.isPresent()) {
+        return range.get().end.line;
+      }
+    }
+    // No import exists, add below package declaration
+    if (tree.getPackageDeclaration().isPresent()) {
+      Optional<Range> range = tree.getPackageDeclaration().get().getRange();
+      if (range.isPresent()) {
+        return range.get().begin.line;
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/Printer.java]: Thank you, that is a good catch, I didn't think about this case as I have never seen it, but we should definitely cover it as well. Fixed and added two unit tests 36df772db26e001ddc3ebea310508afe1df73bad

File: injector/src/main/java/edu/ucr/cs/riple/injector/Printer.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+      }
+    }
+    // No package exists, add import under copy right header if exists, otherwise on the first line.
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/Printer.java]: d911cffce248c144a17789ce90650b32ff6b5c44

File: injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+            "   Field foo5;",
+            "   @Nullable",
+            "   Object bar3(@Nullable Object p1, @Nullable Object p2, Object p3, @Nullable Object p4) {",
+            "       return new Object();",
+            "   }",
+            "   @Nullable",
+            "   Object bar4() {",
+            "       return new Object();",
+            "   }",
+            "}")
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java:92]: I was not able to fully reproduce the example you gave me, Would you please tell me how I can pass an interface as a parameter for annotation, it seems to it is not possible. 

Instead I used the example below:
```java
@Foo(clazz = String.class, value = "Some description"),
private static @Baz Object foo6(
   final @Baz Map<String, @Baz Object> m,
   @Foo(clazz = String.class, value = "Some argument") @Baz Object o1,
   Object o2,
   Object o3,
   @Foo(clazz = Object.class, value = \"Some other argument\") @Baz Object o4) {
 return new @Baz Runnable() {
   public void run() {
     System.out.print(\"log\");
   }
 };
}
```

Added 5 unit tests on method and parameters and all tests pass. 509aaae3d17204c92a4573aec5094b41669777dd

File: injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java [hunk 16]
```diff
@@ -58,23 +63,29 @@ protected void fillJsonInformation(JSONObject res) {
   }
 
   @Override
-  protected boolean applyToMember(NodeList<BodyDeclaration<?>> clazz, Change change) {
-    final boolean[] success = {false};
+  protected Modification applyToMember(NodeList<BodyDeclaration<?>> clazz, Change change) {
+    final AtomicReference<Modification> ans = new AtomicReference<>();
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java:67]: The reason for using `AtomicReference` is to avoid making a single element array in this case. Please see [comment](https://github.com/ucr-riple/NullAwayAnnotator/pull/90#discussion_r1014872217)

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Modification.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+
+  /** Starting position where the modification should be applied in the source file. */
+  public final Position startPosition;
+  /** Content of modification. */
+  public final String content;
+  /** {@link ElementKind} of the element where this modification is created for. */
+  public final ElementKind kind;
+
+  public Modification(String content, Position startPosition, ElementKind kind) {
+    // Position in javaparser is not 0 indexed and line and column fields are final.
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Modification.java:46]: 6f18d12b2a47ddeafb145d4bbf74604a03d88c04

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+
+/** Represents a deletion of a content in the source file. */
+public class Deletion extends Modification {
+
+  /** The end position of content which should be removed. */
+  private final Position endPosition;
+
+  public Deletion(String content, Position startPosition, Position endPosition, ElementKind kind) {
+    super(content, startPosition, kind);
+    this.endPosition = new Position(endPosition.line - 1, endPosition.column - 1);
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java:40]: 6f18d12b2a47ddeafb145d4bbf74604a03d88c04

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java (hunk 6 above)
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java:51]: Yes, we do not support deletion of annotations written in multiple lines. It wasn't also possible to make such requests from Injector, as we can only make `RemoveAnnotation` instance for marker annotations. This check was implicitly applied, however, I added the precondition explicitly 970f82ed397473080751e5d8051f769645b72f53.

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Replacement.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+/** Represents a replacement of a content in the source file. */
+public class Replacement extends Modification {
+
+  /** The end position of the content that should be replaced. */
+  private final Position endPosition;
+
+  public Replacement(
+      String content, Position startPosition, Position endPosition, ElementKind kind) {
+    super(content, startPosition, kind);
+    this.endPosition = new Position(endPosition.line - 1, endPosition.column - 1);
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Replacement.java:41]: 6f18d12b2a47ddeafb145d4bbf74604a03d88c04

File: injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (msridhar) [reply injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java:67]: @lazaroclapp this is just to create a "mutable" local that can be updated inside the nested lambda.  `AtomicReference` is just a common way I've seen to do this; it has nothing to do with concurrency.  If there were a `MutableRef<T>` type we could just use that (but I don't know of one in the standard library, other than something like `WeakReference<T>` which would be even weirder to use).

File: injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java [hunk 20]
```diff
@@ -58,23 +63,29 @@ protected void fillJsonInformation(JSONObject res) {
...
+                      Optional<Range> range = fieldDeclaration.getRange();
+                      range.ifPresent(
+                          value ->
+                              ans.set(change.visit(ElementKind.FIELD, fieldDeclaration, value)));
                       break;
                     }
                   }
                 }));
-    return success[0];
+    return ans.get();
```
[the remark this answers is not in the document]
Comment (msridhar) [reply injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java:88]: We already treat `AtomicReference.get()` as `@Nullable`:

https://github.com/uber/NullAway/blob/19bbb915357a323dbb0b449449520f360b237221/nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java#L628

But I think that NullAway is not running on this repo, is that right, @nimakarimipour?  We should probably add NullAway (and Error Prone) checking at some point soon (but not right now).l

File: injector/src/main/java/edu/ucr/cs/riple/injector/Printer.java [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+  /**
+   * Computes the offset where the import declarations should be added. Below is the steps it takes
+   * to compute the correct offset. At each step, if the desired location is not found, it goes to
+   * following step.
+   *
+   * <ol>
+   *   <li>Immediately after the last existing import statement
+   *   <li>Immediately after the package declaration statement
+   *   <li>Immediately after the copy right header
+   *   <li>Immediately before non-empty line
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/Printer.java]: d0e535731d5221e0bd0cedbc4ffd9761bfff8cba

File: injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java:88]: @msridhar Yes, we are not running NullAway on this repo, but we do run ErrorProne on all packages. I add `@Nullable` annotations to get warnings from IDE for now and make the migration to NullAway in future easier.

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+    StringBuilder line = new StringBuilder(lines.get(startPosition.line));
+    line.delete(start, end + 1);
+    // char at start is removed, head is one character before start.
+    int head = start - 1;
+    // if head is surrounded by white space, remove the extra whitespace.
+    // (  ...) -> ( ...)
+    if (head > 1 && line.charAt(head - 1) == ' ' && line.charAt(head) == ' ') {
+      line.deleteCharAt(head);
+    }
+    // if head is not alphabetic and is followed by a whitespace, remove it.
```
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java]: Yes,  Actually just the check for non-alphanumerical is enough for all cases. The reason for this check it is that we would like to remove the space inserted by additions on deletions. Please see the example below:
```java
void foo(Object bar)
```
If we make `bar` `@Nullable`, we get:
```java
void foo(@Nullable Object bar)
```

If we just delete the annotation later we will leave the empty space inserted by addition and produce output below:
```java
void foo( Object bar)
```

Please note we cannot always delete the space in between, consider the example below:
```java
public@annot Object foo();
```
If always delete the space we will mix two tokens and produce `publicObject foo();`, therefore we should only remove the space if there exists a non-alphanumerical to keep tokens separated.

Added a test case as well: 7bfb20a65d0e9ebd3b48800e464ed0059f1021c7

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Replacement.java (hunk 7 above)
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Replacement.java:53]: Thank you very much for catching this bug. I updated the replacement logic according to `core` module needs 6e751264bc5145678df434ca122bf68939ca729a. I don't think we plan on adding annotations with argument on parameters for now. I added the precondition so we can get notified if such annotations are ever requested from `core` module so we can enhance it.

**Comment by nimakarimipour** [comment]
@lazaroclapp Thank you very much for the review, this is ready for another round of review.

File: injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (lazaroclapp) [reply injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java:92]: > Would you please tell me how I can pass an interface as a parameter for annotation, it seems to it is not possible.

Not sure I understand the question. I think all you need is an annotation with `value` of type `Object` or `Runnable` to pass an anonymous subclass of `Runnable` to it. But, also, that probably isn't the most interesting part of my convoluted example, so it's fine to leave it out. I would, however, make sure `o1` and `o2` are on the same line, and similarly for `o3` and `o4`, to make sure we are testing the "column" position info in addition to the "line" position info. So:

```
@Foo(clazz = String.class, value = "Some description"),
private static @Baz Object foo6(
   final @Baz Map<String, @Baz Object> m,
   @Foo(clazz = String.class, value = "Some argument") @Baz Object o1, Object o2,
   Object o3, @Foo(clazz = Object.class, value = \"Some other argument\") @Baz Object o4) {
 return new @Baz Runnable() {
   public void run() {
     System.out.print(\"log\");
   }
 };
}
```

(While weird - and un-GJF'ed - formatting, that should test the injector a bit more)

File: injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (lazaroclapp) [reply injector/src/main/java/edu/ucr/cs/riple/injector/location/OnField.java:67]: Ok, wasn't familiar with the idiom, but makes sense to me 🙂

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (lazaroclapp) [reply injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java]: > Yes, Actually just the check for non-alphanumerical is enough for all cases. 

I don't think that's correct. Consider this horrifying but JLS-valid annotation name:

```
@interface AnnotationWithNonAlphanumericSuffix$$$$_ { }

class Foo {
    public void foo(@AnnotationWithNonAlphanumericSuffix$$$$_@Nullable Object o) { }
}
```

I suspect removing `@Nullable` there will give you broken code. The right check is probably `Character.isJavaIdentifierPart(...)`, rather than `Character.isLetterOrDigit(...)`... but the simpler and most robust thing to do would probably be just `Character.isWhitespace(...)` and drop the negation.

File: injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java:76]: Added comment 025f0232a5efe90bc035033e624dcc62f9cffd1e

File: injector/src/test/java/edu/ucr/cs/riple/injector/OnParameterInjectionTest.java [hunk 23]
```diff
@@ -274,4 +274,204 @@ public void parameter_nullable_with_annotation() {
...
+            "   }",
+            "}")
+        .addChanges(
+            new AddMarkerAnnotation(
+                new OnParameter(
+                    "Main.java",
+                    "edu.ucr.Main",
+                    "foo6(Map<String, Object>, Object, Object, Object, Object)",
+                    4),
+                "javax.annotation.Nullable"))
```
Comment (lazaroclapp) [inline injector/src/test/java/edu/ucr/cs/riple/injector/OnParameterInjectionTest.java:474]: Just saw all these tests! Nice. Never mind my comment about missing the checks for the arguments `o2` and `o4`. I still think it's worth checking it works if they are on the same line as the other two arguments.

**Review by lazaroclapp — COMMENTED** [review]
Made some follow up comments in the discussion threads above when looking at it commit by commit, but otherwise these changes look good to me 🙂

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java]: Hi @lazaroclapp, thank you for catching this bug, yes the current implementation fails with the test case you provided. I changed it to using `isJavaIdentifierPart(...)` and it worked along all other test cases.

`Character.isWhitespace(...)` does not work as it does not remove the left white space in addition. Please see the example below for test case `edu.ucr.cs.riple.injector.RemovalTest > remove_annot_param`:
For `@Nullable Object test(@Nullable Object o)` 
we get:   `@Nullable Object test( Object o)`
and not: `@Nullable Object test(Object o)`

Please let me know if I am missing something. Thank you. 


Also added a test case that only passes if we use `Character.isJavaIdentifierPart(...)`. eed362882eade8a4046e7530237ed1e859308237

File: injector/src/test/java/edu/ucr/cs/riple/injector/OnParameterInjectionTest.java (hunk 23 above)
Comment (nimakarimipour) [reply injector/src/test/java/edu/ucr/cs/riple/injector/OnParameterInjectionTest.java:763]: FYI, I removed some of the redundant test cases. I 9ffe4939c994252e0198a25b1178aa9bdc138eac.

File: injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (lazaroclapp) [reply injector/src/test/java/edu/ucr/cs/riple/injector/ComprehensiveTest.java:94]: Interesting. You are right, can't be an arbitrary anonymous class, but it can be e.g. "an annotation type". Wonder if an anonymous subtype of an enum or an annotation is a thing... but either way, the example should be fine as is 🙂

**Review by lazaroclapp — APPROVED** [review]
LGTM!

File: injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java]: @msridhar @lazaroclapp Actually I don't think we should rely on GJF too much and should do the fixing on our end as much as possible. The reason is that during the process of measuring effects of fix trees (at least for now) we operate multiple pairs of `@Nullable` addition / deletion on one element at each depth. If our consecutive addition / deletion leave a space, in the end wight end up with a method like below:

```java
void foo(                 Object f,                   Object c)
```

Although this is fixable by GJF (which we do not run it within iterations) this can be the end state of source files at iteration `j` and the input for iteration `j+1`. To make sure we will not run the parser on complicated cases within iterations, it is better to not rely on GJF and fix it ourself.
